spynnaker/pyNN/models/neural_projections/connectors/csa_connector.py

```diff
@@ -0,0 +1,159 @@
+from spinn_utilities.overrides import overrides
+from .abstract_connector import AbstractConnector
+
+import logging
+import numpy
+import csa
+
+logger = logging.getLogger(__name__)
+
+
+class CSAConnector(AbstractConnector):
+    """ Make connections using a Connection Set Algebra (Djurfeldt 2012)
+        description between the neurons in the pre- and post-populations.
+    """
+
+    __slots = [
+        "_cset"]
+
+    def __init__(
+            self, cset,
+            safe=True, callback=None, verbose=False):
+        """
+
+        :param '?' cset:
+            A description of the connection set between populations
+        """
+        super(CSAConnector, self).__init__(safe, verbose)
+        self._cset = cset
+        self._verbose = verbose
+
+        # think this is probably needed here
+        self._pair_list = None
+        self._full_connection_set = None
+        self._full_cset = None
+
+    @overrides(AbstractConnector.get_delay_maximum)
+    def get_delay_maximum(self):
+        n_connections_max = self._n_pre_neurons * self._n_post_neurons
+        # we can probably look at the array and do better than this?
+        return self._get_delay_maximum(
+            self._delays, n_connections_max)
+
+    @overrides(AbstractConnector.get_delay_variance)
+    def get_delay_variance(
+            self, pre_slices, pre_slice_index, post_slices,
+            post_slice_index, pre_vertex_slice, post_vertex_slice):
+        return self._get_delay_variance(self._delays, None)
+
+    def _get_n_connections(self, pre_vertex_slice, post_vertex_slice):
+        # do the work from self._cset in here
+        # get the values for this slice
+        pre_lo = pre_vertex_slice.lo_atom
+        pre_hi = pre_vertex_slice.hi_atom
+        post_lo = post_vertex_slice.lo_atom
+        post_hi = post_vertex_slice.hi_atom
+
+        # this is where the magic needs to happen somehow
+        if self._full_cset is None:
+            self._full_cset = [x for x in csa.cross(
+                range(self._n_pre_neurons),
+                range(self._n_post_neurons)) * self._cset]
+
+        # use CSA to cross the range of this vertex's neurons with the cset
+        self._pair_list = csa.cross(
+            range(pre_lo, pre_hi+1),
+            range(post_lo, post_hi+1)) * self._full_cset
+
+        # Get the lists of pre and post neurons that result from this
+        self._pre_neurons = [x[0] for x in self._pair_list]
+        self._post_neurons = [x[1] for x in self._pair_list]
```

Comment:
I think this and self._pre_neurons could be returned rather than stored as they are calculated at each call

Reply:
Don't even need to be returned as they can be obtained from pair_list whenever required - done.